main/src/ca/uwaterloo/flix/language/phase/jvm/BackendType.scala

```diff
@@ -44,63 +46,60 @@ object VoidableType {
   */
 sealed trait BackendType extends VoidableType {
   def toDescriptor: String = this match {
-    case BackendType.Bool => "Z"
-    case BackendType.Char => "C"
-    case BackendType.Int8 => "B"
-    case BackendType.Int16 => "S"
-    case BackendType.Int32 => "I"
-    case BackendType.Int64 => "J"
-    case BackendType.Float32 => "F"
-    case BackendType.Float64 => "D"
+    case BackendType.Primitive(tpe) => tpe.toDescriptor
     case BackendType.Array(tpe) => s"[${tpe.toDescriptor}"
     case BackendType.Reference(ref) => ref.toDescriptor
   }
 
   /**
     * Returns the erased type, either itself if `this` is primitive or `java.lang.Object`
     * if `this` is an array or a reference.
+    *
     * @return
     */
   def toErased: BackendType = this match {
-    case BackendType.Bool | BackendType.Char | BackendType.Int8 | BackendType.Int16 |
-         BackendType.Int32 | BackendType.Int64 | BackendType.Float32 | BackendType.Float64 => this
+    case BackendType.Primitive(_) => this
     case BackendType.Array(_) | BackendType.Reference(_) => BackendObjType.JavaObject.toTpe
   }
 
   /**
     * A string representing the erased type. This is used for parametrized class names.
     */
   val toErasedString: String = this match {
-    case BackendType.Bool => "Bool"
-    case BackendType.Char => "Char"
-    case BackendType.Int8 => "Int8"
-    case BackendType.Int16 => "Int16"
-    case BackendType.Int32 => "Int32"
-    case BackendType.Int64 => "Int64"
-    case BackendType.Float32 => "Float32"
-    case BackendType.Float64 => "Float64"
+    case BackendType.Primitive(tpe) => tpe.toErasedString
     case BackendType.Array(_) | BackendType.Reference(_) => "Obj"
   }
-}
-
-object BackendType {
-  case object Bool extends BackendType
 
-  case object Char extends BackendType
+  def is64BitWidth: Boolean = this match {
+    case BackendType.Primitive(PrimitiveType.Int64) => true
+    case BackendType.Primitive(PrimitiveType.Float64) => true
+    case BackendType.Primitive(_) | BackendType.Array(_) |
+         BackendType.Reference(_) => false
+  }
 
-  case object Int8 extends BackendType
+  /**
+    * Returns the Array fill type for the value of the type specified by `tpe`
+    */
+  def toArrayFillType: String = s"([${this.toErased.toDescriptor}${this.toErased.toDescriptor})V"
 
-  case object Int16 extends BackendType
+}
 
-  case object Int32 extends BackendType
+object BackendType {
 
-  case object Int64 extends BackendType
+  case class Primitive(tpe: PrimitiveType) extends BackendType
```

Comment:
This is very Flix like but maybe `Primitive` should just be a marker interface, i.e. a trait that is implemented by the primitive types. Those primitive types would look exactly like they did before (so still accessed by `BackendType.Int32`). This would all code "ignore" the fact that this trait is there but you can match on it if you want to. Hope that makes sense